In [36]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from path import Path
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [37]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    Path("Resources/crypto_market_data.csv"),
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [38]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [39]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

In [40]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
scaled_data = StandardScaler().fit_transform(df_market_data)

In [41]:
# Create a DataFrame with the scaled data
df_market_data_scaled = pd.DataFrame(
    scaled_data,
    columns=df_market_data.columns
)

# Copy the crypto names from the original data
df_market_data_scaled["coin_id"] = df_market_data.index

# Set the coinid column as index
df_market_data_scaled = df_market_data_scaled.set_index("coin_id")

# Display sample data
df_market_data_scaled.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

In [42]:
# Initialize the K-Means model with four clusters
model = KMeans(n_clusters=4)

In [43]:
# Fit the K-Means model using the scaled data
model.fit(df_market_data_scaled)

KMeans(n_clusters=4)

In [44]:
# Predict the clusters to group the cryptocurrencies using the scaled data
crypto_clusters_k4 = model.predict(df_market_data_scaled)

# View the resulting array of cluster values.
print(crypto_clusters_k4)

[3 3 0 0 3 3 3 3 3 0 0 0 0 3 0 3 0 0 3 0 0 3 0 0 0 0 0 0 3 0 0 0 1 3 0 0 2
 0 0 0 0]


In [45]:
# Add a new column to the DataFrame with the predicted clusters with k=4
df_market_data_scaled["crypto_cluster_k4"] = crypto_clusters_k4

# Display sample data
df_market_data_scaled.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  crypto_cluster_k4  
coin_id                                                      
bitcoin                        -0.251637                  3  
ethereum                       -0.199352                  3  
tether                         -0.282061                  0  
ripple                         -0.295546                  0  
bitcoin-cash                   -0.270317                  3

In [46]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_14d"` and `y="price_change_percentage_1y"`.
# Group the results by the clusters using `by="crypto_cluster_k4".
# Set the hover to the coin id using `hover_cols=["coin_id"]`.

df_market_data_scaled.hvplot.scatter(
    x="price_change_percentage_14d",
    y="price_change_percentage_1y",
    by="crypto_cluster_k4",
    hover_cols=["coin_id"],
    marker=["hex", "square", "cross", "inverted_triangle", "triangle"],
)

:NdOverlay   [crypto_cluster_k4]
   :Scatter   [price_change_percentage_14d]   (price_change_percentage_1y,coin_id)

---

In [47]:
# Create a list with the number of k-values to try
# Use a range from 1 to 11
k = list(range(1, 11))

In [48]:
# Create an empy list to store the inertia values
inertia = []

In [49]:
# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inirtia list
for i in k:
    model = KMeans(n_clusters=i, random_state=0)
    model.fit(df_market_data_scaled)
    inertia.append(model.inertia_)

In [50]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {
    "k": k,
    "inertia": inertia
}

# Create a DataFrame with the data to plot the Elbow curve
df_elbow = pd.DataFrame(elbow_data)

In [51]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
df_elbow.hvplot(x='k', y='inertia', title='Elbow Curve', xticks=k)

:Curve   [k]   (inertia)

**Question:** What is the best value for `k`?

**Answer:** Based on the elbow curve, the best value for k is 5.

---

In [52]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

In [53]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
market_pca_data = pca.fit_transform(df_market_data_scaled)

# View the first five rows of the DataFrame. 
market_pca_data[:5]

array([[ 2.05913887, -0.53668446, -0.43823774],
       [ 1.95640089, -0.41082149, -1.0169655 ],
       [-0.99014153, -0.45422783,  0.69002131],
       [-0.9731886 , -0.49464992,  0.55123402],
       [ 3.33036692, -1.02231141, -0.50134143]])

In [54]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
pca.explained_variance_ratio_

array([0.38939231, 0.29166039, 0.20810061])

**Question** What is the total explained variance of the three principal components?

**Answer** 89%. To find the total explained variance we add up the principal components, the first has 36.94% of the variance, the second has 29.17%, the third has 22.89%, which sum up to 89%. 

In [55]:
# Creating a DataFrame with the PCA data
df_market_data_pca = pd.DataFrame(
    market_pca_data, 
    columns=["PC1", "PC2", "PC3"]
)

# Copy the crypto names from the original data
df_market_data_pca["coin_id"] = df_market_data.index

# Set the coinid column as index
df_market_data_pca = df_market_data_pca.set_index("coin_id")

# Display sample data
df_market_data_pca.head()

PC1       PC2       PC3
coin_id                                   
bitcoin       2.059139 -0.536684 -0.438238
ethereum      1.956401 -0.410821 -1.016966
tether       -0.990142 -0.454228  0.690021
ripple       -0.973189 -0.494650  0.551234
bitcoin-cash  3.330367 -1.022311 -0.501341

In [61]:
# Initiate a new  K-Means algorithm using the PCA DataFrame to group 
# the cryptocurrencies. Set the `n_clusters` parameter equal to 
# the best value for `k` found before. View the resulting array. 

# Initialize the K-Means model
model = KMeans(n_clusters=5)

# Fit the model
model.fit(df_market_data_pca)

# Predict clusters
crypto_clusters_k5 = model.predict(df_market_data_pca)

# View the resulting array
crypto_clusters_k5

array([1, 1, 1, 1, 1, 1, 4, 4, 1, 1, 4, 1, 1, 4, 1, 4, 4, 1, 1, 1, 4, 1,
       1, 1, 4, 1, 1, 4, 1, 3, 1, 1, 2, 1, 3, 4, 0, 1, 4, 1, 4],
      dtype=int32)

In [62]:
# - From the original DataFrame, add the `price_change_percentage_1y` and `price_change_percentage_14d` columns.
# - Add a column with the predicted cluster values identified using a k value of 4. (The predicted cluster values were calculated in the “Cluster Cryptocurrencies with K-means” section.)
# - Add a column with the predicted cluster values identified using the optimal value for k. 


# Add the price_change_percentage_1y column from the original data
df_market_data_pca["price_change_percentage_1y"] = df_market_data["price_change_percentage_1y"]

# Add the price_change_percentage_14d column from the original data
df_market_data_pca["price_change_percentage_14d"] = df_market_data["price_change_percentage_14d"]

# Add a new column to the DataFrame with the predicted clusters using the best value of k
df_market_data_pca["crypto_cluster_k5"] = crypto_clusters_k5

# Add a new column to the DataFrame with the predicted clusters using k=4
df_market_data_pca["crypto_cluster_k4"] = crypto_clusters_k4

# Display sample data
df_market_data_pca.head()

PC1       PC2       PC3  price_change_percentage_1y  \
coin_id                                                                  
bitcoin       2.059139 -0.536684 -0.438238                    37.51761   
ethereum      1.956401 -0.410821 -1.016966                   101.96023   
tether       -0.990142 -0.454228  0.690021                     0.01954   
ripple       -0.973189 -0.494650  0.551234                   -16.60193   
bitcoin-cash  3.330367 -1.022311 -0.501341                    14.49384   

              price_change_percentage_14d  crypto_cluster_k5  \
coin_id                                                        
bitcoin                           6.57509                  1   
ethereum                          4.80849                  1   
tether                            0.00640                  1   
ripple                            2.24984                  1   
bitcoin-cash                     14.75334                  1   

              crypto_cluster_k4  
coin_id                          
bitcoin                       3  
ethereum                      3  
tether                        0  
ripple                        0  
bitcoin-cash                  3

---

In [63]:
# Create a scatter plot for the Crypto Clusters using k=4 data.
# Use the PCA data to create a scatter plot with hvPlot by setting 
# x="price_change_percentage_14d" and y="price_change_percentage_1y".
# Group by the clusters using `by="crypto_cluster_k4".
# Set the hover colors to the coin id with `hover_cols=["coin_id"]
# Create a descriptive title for the plot using the title parameter.
scatter_plot_k4 = df_market_data_pca.hvplot.scatter(
    x="price_change_percentage_14d",
    y="price_change_percentage_1y",
    by="crypto_cluster_k4",
    hover_cols=["coin_id"],
    title='Clustered k=4 Market Data'
)
scatter_plot_k4

:NdOverlay   [crypto_cluster_k4]
   :Scatter   [price_change_percentage_14d]   (price_change_percentage_1y,coin_id)

In [64]:
# Create a scatter plot for the Crypto Clusters using k=5 data.
# Use the PCA data to create a scatter plot with hvPlot by setting 
# x="price_change_percentage_14d" and y="price_change_percentage_1y".
# Group by the clusters using `by="crypto_cluster_k5".
# Set the hover colors to the coin id with `hover_cols=["coin_id"]
# Create a descriptive title for the plot using the title parameter.
scatter_plot_k5 = df_market_data_pca.hvplot.scatter(
    x="price_change_percentage_14d",
    y="price_change_percentage_1y",
    by="crypto_cluster_k5",
    hover_cols=["coin_id"],
    title='PCA Clustered k=5 Market Data'
)
scatter_plot_k5

:NdOverlay   [crypto_cluster_k5]
   :Scatter   [price_change_percentage_14d]   (price_change_percentage_1y,coin_id)

In [65]:
# Compare both scatter plots
scatter_plot_k4 + scatter_plot_k5

:Layout
   .NdOverlay.I  :NdOverlay   [crypto_cluster_k4]
      :Scatter   [price_change_percentage_14d]   (price_change_percentage_1y,coin_id)
   .NdOverlay.II :NdOverlay   [crypto_cluster_k5]
      :Scatter   [price_change_percentage_14d]   (price_change_percentage_1y,coin_id)

**Question:** What value of `k` seems to create the most accurate clusters to group cryptocurrencies according to their profitability?

**Answer:** Based on the scatter plots, it seems like k = 4 is the most accurate to group cryptocurrencies according to their profitibility 